### ISSUE TITLE
Allow symbols to specify functions

### ISSUE THREAD

**Issue by lgrapenthin** [onyx-platform/onyx#253]
The EDN spec (www.edn-format.org) says
"Keywords are identifiers that typically designate themselves. They are semantically akin to enumeration values. [...]"

and

"Symbols are used to represent identifiers, [...]"

The proposed enhancement would be to allow (and use as a default in documentation and examples) symbols where functions should be specified in data.

Advantages:

- Improved readabilty: More explicit where functions are/should be specified
- IDEs are usually configured to code-jump on symbols, not keywords. (I was thinking about a Cider issue to code jump on keywords but since this pattern is not used elsewhere I doubt they'd accept it)

Disadvantages:
- None, unless you intended to use symbols differently
- Users might forget to quote, but this should be easy to report.

**Comment by MichaelDrogalis** [comment]
A disadvantage of this approach is that, in practice, catalogs are often constructed on the client side of an architecture - sometimes in JavaScript or ClojureScript. These places can't reference symbols to functions because they either aren't on their classpath, or these environments aren't home to a JVM to begin with.

I'll consider widening the schema to allow for either a keyword or a symbol, but a symbol will not be the default. Using a keyword to point to a function is indeed a little unusual, but the circumstances warrant it.

Thanks for the suggestion! I'll mull it over some more.

**Comment by lgrapenthin** [comment]
I don't understand the disadvantage. If a platform can't reference a symbol to a function neither can it reference a keyword to a function. 

To make it more clear I am referring to literal symbols. Intended is a semantic enhancement.

**Comment by MichaelDrogalis** [comment]
Right, I believe I am on the same page with you regarding what the proposed adjustment is. While Onyx catalogs, workflows, and flow conditions are often presented in the EDN format in example code, the end game for Onyx is that these data structures are constructed in non-JVM environments. When they are built up inside a JVM, it is not correct to assume that these functions will exist on your classpath. Onyx takes a strong departure from Storm and Hadoop with this attitude. Submitted jobs have an implicit contract that functions will exist on the classpath when the job begins on cluster machines. The submitter of the job doesn't need to have a reference to these symbols. That's essentially where Onyx gets its deployment flexibility from.

In short, I think for development it would probably be worthwhile to also allow symbol literals, but the default will remain keyword-represented functions.

**Comment by lgrapenthin** [comment]
I am failing to see the requirement that using a literal symbol requires the existence of a local var.

 E. g.

```clojure
{...
 :onyx/type :function
 :onyx/fn 'your.ns/your-fn}
```

**Comment by MichaelDrogalis** [comment]
Ah - apologies. I was picturing `#'your.ms/your-fn`. Right - a symbol literal doesn't require the local var to be present.

I will probably loosen the schema next release to allow for this, but I'm really not interested in making this the default. Thanks though.

**Comment by MichaelDrogalis** [comment]
After some more thought, I've decide to keep the schema for `:onyx/fn` and related strictly as a keyword. This is much more in line with the long term goals of Onyx. Tools can adjust to do code-jumping on namespaced keywords - distorting the information model wouldn't be something we could easily back pedal on. Thanks for your input.
